Submodule temporal 81d3454..308afb3:
diff --git a/temporal/common/persistence/sql/sqlplugin/db_handle.go b/temporal/common/persistence/sql/sqlplugin/db_handle.go
index a4f4071c7..6aa23b2de 100644
--- a/temporal/common/persistence/sql/sqlplugin/db_handle.go
+++ b/temporal/common/persistence/sql/sqlplugin/db_handle.go
@@ -29,6 +29,7 @@ import (
 	"errors"
 	"fmt"
 	"io"
+	"strings"
 	"sync"
 	"sync/atomic"
 	"syscall"
@@ -45,6 +46,7 @@ import (
 const (
 	// TODO: this should be dynamic config.
 	sessionRefreshMinInternal = 1 * time.Second
+	accessTokenExpireErrorMsg = "The access token has expired. Please acquire a new token and retry."
 )
 
 var (
@@ -177,7 +179,8 @@ func (h *DatabaseHandle) ConvertError(err error) error {
 		errors.Is(err, io.EOF) ||
 		errors.Is(err, syscall.ECONNRESET) ||
 		errors.Is(err, syscall.ECONNABORTED) ||
-		errors.Is(err, syscall.ECONNREFUSED) {
+		errors.Is(err, syscall.ECONNREFUSED) ||
+		(err != nil && strings.Contains(err.Error(), accessTokenExpireErrorMsg)) {
 		h.reconnect(true)
 		return serviceerror.NewUnavailable(fmt.Sprintf("database connection lost: %s", err.Error()))
 	}
