diff --git a/docker/auto-setup.sh b/docker/auto-setup.sh
index 461e40f..6a91681 100755
--- a/docker/auto-setup.sh
+++ b/docker/auto-setup.sh
@@ -206,7 +206,8 @@ setup_postgres_schema() {
     SCHEMA_DIR=${TEMPORAL_HOME}/schema/postgresql/${POSTGRES_VERSION_DIR}/temporal/versioned
     # Create database only if its name is different from the user name. Otherwise PostgreSQL container itself will create database.
     if [[ ${DBNAME} != "${POSTGRES_USER}" && ${SKIP_DB_CREATE} != true ]]; then
-        temporal-sql-tool \
+        echo "Creating database ${DBNAME}."
+        until temporal-sql-tool \
             --plugin ${DB} \
             --ep "${POSTGRES_SEEDS}" \
             -u "${POSTGRES_USER}" \
@@ -218,9 +219,15 @@ setup_postgres_schema() {
             --tls-key-file "${POSTGRES_TLS_KEY_FILE}" \
             --tls-ca-file "${POSTGRES_TLS_CA_FILE}" \
             --tls-server-name "${POSTGRES_TLS_SERVER_NAME}" \
-            create
+            create \
+            ; do
+                echo "Failed to create database ${DBNAME}. Retrying..."
+                sleep 10
+        done
     fi
-    temporal-sql-tool \
+
+    echo "Setting up schema for ${DBNAME}."
+    until temporal-sql-tool \
         --plugin ${DB} \
         --ep "${POSTGRES_SEEDS}" \
         -u "${POSTGRES_USER}" \
@@ -232,8 +239,14 @@ setup_postgres_schema() {
         --tls-key-file "${POSTGRES_TLS_KEY_FILE}" \
         --tls-ca-file "${POSTGRES_TLS_CA_FILE}" \
         --tls-server-name "${POSTGRES_TLS_SERVER_NAME}" \
-        setup-schema -v 0.0
-    temporal-sql-tool \
+        setup-schema -v 0.0 \
+        ; do
+            echo "Failed to set up schema ${DBNAME}. Retrying..."
+            sleep 10
+    done
+
+    echo "Updating schema for ${DBNAME}."
+    until temporal-sql-tool \
         --plugin ${DB} \
         --ep "${POSTGRES_SEEDS}" \
         -u "${POSTGRES_USER}" \
@@ -245,13 +258,18 @@ setup_postgres_schema() {
         --tls-key-file "${POSTGRES_TLS_KEY_FILE}" \
         --tls-ca-file "${POSTGRES_TLS_CA_FILE}" \
         --tls-server-name "${POSTGRES_TLS_SERVER_NAME}" \
-        update-schema -d "${SCHEMA_DIR}"
+        update-schema -d "${SCHEMA_DIR}" \
+        ; do
+            echo "Failed to update schema ${DBNAME}. Retrying..."
+            sleep 10
+    done
 
     # Only setup visibility schema if ES is not enabled
     if [[ ${ENABLE_ES} == false ]]; then
       VISIBILITY_SCHEMA_DIR=${TEMPORAL_HOME}/schema/postgresql/${POSTGRES_VERSION_DIR}/visibility/versioned
       if [[ ${VISIBILITY_DBNAME} != "${POSTGRES_USER}" && ${SKIP_DB_CREATE} != true ]]; then
-          temporal-sql-tool \
+          echo "Creating visibility database ${VISIBILITY_DBNAME}."
+          until temporal-sql-tool \
               --plugin ${DB} \
               --ep "${POSTGRES_SEEDS}" \
               -u "${POSTGRES_USER}" \
@@ -263,9 +281,15 @@ setup_postgres_schema() {
               --tls-key-file "${POSTGRES_TLS_KEY_FILE}" \
               --tls-ca-file "${POSTGRES_TLS_CA_FILE}" \
               --tls-server-name "${POSTGRES_TLS_SERVER_NAME}" \
-              create
+              create \
+              ; do
+                  echo "Failed to create visibility database ${VISIBILITY_DBNAME}. Retrying..."
+                  sleep 10
+          done
       fi
-      temporal-sql-tool \
+
+      echo "Setting up schema for visibility database ${VISIBILITY_DBNAME}."
+      until temporal-sql-tool \
           --plugin ${DB} \
           --ep "${POSTGRES_SEEDS}" \
           -u "${POSTGRES_USER}" \
@@ -277,8 +301,14 @@ setup_postgres_schema() {
           --tls-key-file "${POSTGRES_TLS_KEY_FILE}" \
           --tls-ca-file "${POSTGRES_TLS_CA_FILE}" \
           --tls-server-name "${POSTGRES_TLS_SERVER_NAME}" \
-          setup-schema -v 0.0
-      temporal-sql-tool \
+          setup-schema -v 0.0 \
+          ; do
+              echo "Failed to set up schema for visibility database ${VISIBILITY_DBNAME}. Retrying..."
+              sleep 10
+      done
+
+      echo "Updating schema for visibility database ${VISIBILITY_DBNAME}."
+      until temporal-sql-tool \
           --plugin ${DB} \
           --ep "${POSTGRES_SEEDS}" \
           -u "${POSTGRES_USER}" \
@@ -290,7 +320,11 @@ setup_postgres_schema() {
           --tls-key-file "${POSTGRES_TLS_KEY_FILE}" \
           --tls-ca-file "${POSTGRES_TLS_CA_FILE}" \
           --tls-server-name "${POSTGRES_TLS_SERVER_NAME}" \
-          update-schema -d "${VISIBILITY_SCHEMA_DIR}"
+          update-schema -d "${VISIBILITY_SCHEMA_DIR}" \
+          ; do
+              echo "Failed to update schema for visibility database ${VISIBILITY_DBNAME}. Retrying..."
+              sleep 10
+      done
     fi
 }
 
@@ -369,8 +403,11 @@ register_default_namespace() {
     echo "Registering default namespace: ${DEFAULT_NAMESPACE}."
     if ! temporal operator namespace describe "${DEFAULT_NAMESPACE}"; then
         echo "Default namespace ${DEFAULT_NAMESPACE} not found. Creating..."
-        temporal operator namespace create --retention "${DEFAULT_NAMESPACE_RETENTION}" --description "Default namespace for Temporal Server." "${DEFAULT_NAMESPACE}"
-        echo "Default namespace ${DEFAULT_NAMESPACE} registration complete."
+        until temporal operator namespace create --retention "${DEFAULT_NAMESPACE_RETENTION}" --description "Default namespace for Temporal Server." "${DEFAULT_NAMESPACE}"; do
+            echo "Failed to register default namespace ${DEFAULT_NAMESPACE}. Retrying..."
+            sleep 10
+        done
+        echo "Default namespace ${DEFAULT_NAMESPACE} registered."
     else
         echo "Default namespace ${DEFAULT_NAMESPACE} already registered."
     fi
@@ -386,14 +423,19 @@ add_custom_search_attributes() {
     echo "Adding Custom*Field search attributes."
     # TODO: Remove CustomStringField
 # @@@SNIPSTART add-custom-search-attributes-for-testing-command
-    temporal operator search-attribute create --namespace "${DEFAULT_NAMESPACE}" \
+    until temporal operator search-attribute create --namespace "${DEFAULT_NAMESPACE}" \
         --name CustomKeywordField --type Keyword \
         --name CustomStringField --type Text \
         --name CustomTextField --type Text \
         --name CustomIntField --type Int \
         --name CustomDatetimeField --type Datetime \
         --name CustomDoubleField --type Double \
-        --name CustomBoolField --type Bool
+        --name CustomBoolField --type Bool \
+        ; do
+            echo "Failed to add custom search attributes. Retrying..."
+            sleep 10
+    done
+    echo "Custom search attributes added."
 # @@@SNIPEND
 }
 
@@ -421,6 +463,7 @@ if [[ ${SKIP_SCHEMA_SETUP} != true ]]; then
     validate_db_env
     wait_for_db
     setup_schema
+    echo "setup_schema completed."
 fi
 
 if [[ ${ENABLE_ES} == true ]]; then
@@ -429,5 +472,18 @@ if [[ ${ENABLE_ES} == true ]]; then
     setup_es_index
 fi
 
-# Run this func in parallel process. It will wait for server to start and then run required steps.
-setup_server &
+if [[ ${WAIT_FOR_SETUP_SERVER} == true ]]; then
+    # If running as k8s job, wait for setup_server to complete.
+    echo "Waiting for setup_server to complete"
+    setup_server
+    echo "setup_server completed."
+
+    echo "Terminating istio proxy"
+    curl -s -XPOST http://localhost:15020/quitquitquit
+    sleep 10
+else
+    # Run this func in parallel process. It will wait for server to start and then run required steps.
+    echo "Running setup_server in parallel."
+    setup_server &
+fi
+echo "Exiting auto-setup"
